File: scoring/rules.py
"""
scoring/rules.py — חוקי הניקוד כדאטה.

הכלל החשוב ביותר בארכיטקטורה הזאת:
    ליגה חדשה או שינוי חוקים = שורה חדשה ב-game.scoring_rulesets.
    לא commit. לא deploy. לא if league == 'IL'.

כל RuleSet נושא version. ניקוד היסטורי נשמר עם ה-version שבו חושב,
כך ששינוי חוקים באמצע העונה לא משנה תוצאות עבר.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Mapping

from .models import Position

# מכפילי שער לפי עמדה — שוער שמבקיע שווה יותר מחלוץ. קלאסיקה של פנטזי.
DEFAULT_GOAL_POINTS: dict[Position, float] = {"GK": 10, "DEF": 8, "MID": 6, "FWD": 5}
DEFAULT_CLEAN_SHEET: dict[Position, float] = {"GK": 5, "DEF": 4, "MID": 1, "FWD": 0}
DEFAULT_CONCEDED: dict[Position, float] = {"GK": -1, "DEF": -1, "MID": 0, "FWD": 0}


@dataclass(frozen=True, slots=True)
class PersonalRules:
    goal: Mapping[Position, float] = field(default_factory=lambda: dict(DEFAULT_GOAL_POINTS))
    assist: float = 3
    clean_sheet: Mapping[Position, float] = field(default_factory=lambda: dict(DEFAULT_CLEAN_SHEET))
    clean_sheet_min_minutes: int = 60
    minutes_played: float = 1          # על עצם ההשתתפות
    minutes_60_plus: float = 2         # במקום, לא בנוסף
    yellow_card: float = -1
    red_card: float = -3
    own_goal: float = -2
    saves_per: int = 3                 # כל 3 הצלות
    saves_points: float = 1
    penalty_saved: float = 5
    penalty_missed: float = -2
    goals_conceded_per: int = 2
    goals_conceded: Mapping[Position, float] = field(default_factory=lambda: dict(DEFAULT_CONCEDED))


@dataclass(frozen=True, slots=True)
class ResultBonusRules:
    """בונוס תוצאה — ניתן פר שחקן, על סמך תוצאת הקבוצה האמיתית שלו."""
    win: float = 4
    draw: float = 1
    loss: float = 0
    require_minutes: int = 0     # 0 = גם ספסלן מקבל; 1+ = רק מי ששיחק


@dataclass(frozen=True, slots=True)
class VirtualGoalRules:
    """
    "כל 2 שערים קבוצתיים במציאות = שער וירטואלי (+5)".

    aggregation קובע איך סופרים, וזו החלטת מוצר אמיתית:
      · 'pooled'   – סוכמים את כל שערי הקבוצות שבהרכב לקופה אחת, ומחלקים ב-2.
                     נדיב, מייצר הרבה נקודות, מתגמל בחירת קבוצות מתקפיות.
      · 'per_team' – כל קבוצה נספרת לחוד, שאריות לא מצטברות.
                     קמצני יותר, מתגמל בחירה של קבוצה שכבשה 2+ בבודדת.
    ברירת המחדל היא pooled כי היא מייצרת עקומת ניקוד חלקה יותר בין משתמשים.
    """
    goals_per_virtual: int = 2
    points: float = 5
    aggregation: str = "pooled"        # 'pooled' | 'per_team'
    count_only_starters: bool = True


@dataclass(frozen=True, slots=True)
class ConstraintRules:
    lineup_size: int = 11
    max_players_per_team: int = 1      # ★ האילוץ המרכזי של המשחק
    bench_size: int = 0
    formation_allowed: tuple[str, ...] = ("4-3-3", "4-4-2", "3-5-2", "4-2-3-1", "5-3-2", "3-4-3")
    require_captain: bool = True


@dataclass(frozen=True, slots=True)
class RuleSet:
    version: int = 1
    league_code: str | None = None      # None = ברירת מחדל גלובלית
    personal: PersonalRules = field(default_factory=PersonalRules)
    result_bonus: ResultBonusRules = field(default_factory=ResultBonusRules)
    virtual_goal: VirtualGoalRules = field(default_factory=VirtualGoalRules)
    constraints: ConstraintRules = field(default_factory=ConstraintRules)
    captain_multiplier: float = 2.0

    # ---- טעינה מ-JSONB של game.scoring_rulesets ------------------------
    @classmethod
    def from_jsonb(cls, data: Mapping[str, Any], *, version: int = 1,
                   league_code: str | None = None) -> "RuleSet":
        p = data.get("personal", {})
        r = data.get("result_bonus", {})
        v = data.get("virtual_goal", {})
        c = data.get("constraints", {})
        return cls(
            version=version,
            league_code=league_code,
            personal=PersonalRules(
                goal={**DEFAULT_GOAL_POINTS, **(p.get("goal") or {})},
                assist=p.get("assist", 3),
                clean_sheet={**DEFAULT_CLEAN_SHEET, **(p.get("clean_sheet") or {})},
                clean_sheet_min_minutes=p.get("clean_sheet_min_minutes", 60),
                minutes_played=p.get("minutes_played", 1),
                minutes_60_plus=p.get("minutes_60", 2),
                yellow_card=p.get("yellow", -1),
                red_card=p.get("red", -3),
                own_goal=p.get("own_goal", -2),
                saves_per=p.get("save_per", 3),
                saves_points=p.get("save_points", 1),
                penalty_saved=p.get("penalty_saved", 5),
                penalty_missed=p.get("penalty_missed", -2),
                goals_conceded_per=p.get("goals_conceded_per", 2),
                goals_conceded={**DEFAULT_CONCEDED, **(p.get("goals_conceded") or {})},
            ),
            result_bonus=ResultBonusRules(
                win=r.get("W", 4), draw=r.get("D", 1), loss=r.get("L", 0),
                require_minutes=r.get("require_minutes", 0),
            ),
            virtual_goal=VirtualGoalRules(
                goals_per_virtual=v.get("team_goals_per_virtual", 2),
                points=v.get("points", 5),
                aggregation=v.get("aggregation", "pooled"),
                count_only_starters=v.get("count_only_starters", True),
            ),
            constraints=ConstraintRules(
                lineup_size=c.get("lineup_size", 11),
                max_players_per_team=c.get("max_players_per_team", 1),
                bench_size=c.get("bench_size", 0),
                formation_allowed=tuple(c.get("formation_allowed",
                                              ConstraintRules.formation_allowed)),
                require_captain=c.get("require_captain", True),
            ),
            captain_multiplier=data.get("captain_multiplier", 2.0),
        )


# ליגת העל הישראלית — ברירת המחדל של המוצר.
ISRAELI_PREMIER = RuleSet(version=1, league_code="IL_PREMIER")

# דוגמה להרחבה עתידית: ליגה זרה עם 15 שחקנים ועד 2 מאותה קבוצה.
# אין כאן שום קוד חדש — רק ערכים.
EXAMPLE_FOREIGN = RuleSet(
    version=1,
    league_code="EN_PL",
    constraints=ConstraintRules(lineup_size=15, max_players_per_team=2, bench_size=4),
    virtual_goal=VirtualGoalRules(goals_per_virtual=3, points=6, aggregation="per_team"),
)

File: supabase/functions/dubid-score-gameweek/_lib/scoring/rules.ts
/**
 * ⚠ נוצר אוטומטית — אל תערכו כאן.
 *
 * הועתק מ-src/lib/ על ידי scripts/sync-edge-shared.mjs.
 * כל שינוי כאן יימחק בפריסה הבאה. ערכו את המקור.
 */
/**
 * scoring/rules.ts — חוקי הניקוד כדאטה.
 *
 * ליגה חדשה או שינוי איזון = שורה חדשה בטבלת scoring_rulesets.
 * לא commit, לא deploy, ולא `if (league === 'IL')` בשום מקום.
 */
import type { Position } from './types.ts';

export interface PersonalRules {
  goal: Record<Position, number>;
  assist: number;
  cleanSheet: Record<Position, number>;
  cleanSheetMinMinutes: number;
  minutesPlayed: number;
  minutes60Plus: number;
  yellowCard: number;
  redCard: number;
  ownGoal: number;
  savesPer: number;
  savesPoints: number;
  penaltySaved: number;
  penaltyMissed: number;
  goalsConcededPer: number;
  goalsConceded: Record<Position, number>;
}

export interface ResultBonusRules {
  W: number;
  D: number;
  L: number;
  /** 0 = גם מי שלא שיחק מקבל. 1+ = רק מי ששיחק לפחות כך דקות. */
  requireMinutes: number;
}

export interface VirtualGoalRules {
  goalsPerVirtual: number;
  points: number;
  /**
   * 'pooled'   – כל שערי 11 הקבוצות לקופה אחת, חלוקה אחת. שאריות נשמרות.
   * 'perTeam'  – כל קבוצה לחוד, שאריות נזרקות.
   * ההבדל הוא ~35 נק׳ מול ~12 נק׳ למחזור ממוצע. זו החלטת איזון מוצר.
   */
  aggregation: 'pooled' | 'perTeam';
}

/**
 * ★ הקפטן הדובידי
 *
 * שלושה מנגנונים, וכל אחד מהם קיים כדי לפתור בעיה אמיתית:
 *
 *  1. multiplier ×3 — הימור גדול. פי 2 זה סטנדרט משעמם; פי 3 הופך את
 *     בחירת הקפטן להחלטה המשמעותית ביותר במחזור.
 *
 *  2. חסינות כרטיסים — הכפלה פי 3 של עונש היא רצחנית. לכן הקפטן חסין
 *     לכרטיסים צהובים כשהוא "הרוויח" את זה: שיחק לפחות X דקות
 *     והקבוצה שלו לא הפסידה. כרטיס אדום אף פעם לא נסלח — אחרת
 *     המשתמשים ילמדו לבחור קפטן אגרסיבי ולא קפטן טוב.
 *
 *  3. העברה לסגן — קפטן שלא ירד למגרש הורס מחזור שלם בגלל מידע
 *     שלמשתמש לא היה בזמן הנעילה. במקרה כזה הכפולה עוברת לסגן.
 *     בלי המנגנון הזה, פי 3 הוא הימור לא הוגן ולא כיף.
 *
 *  applyToNegative=false: אם הציון האישי של הקפטן שלילי, לא מכפילים.
 *  ההימור מגדיל רווח, לא הפסד.
 */
export interface DubidCaptainRules {
  multiplier: number;
  /**
   * על מה חל המכפיל:
   *  'personal'        – רק פעולות אישיות (שערים, בישולים, קלין-שיט).
   *  'personalAndResult' – גם בונוס התוצאה הקבוצתית (ברירת מחדל).
   * ברירת המחדל מחזקת את ה-DNA של המשחק: לא מספיק לבחור שחקן טוב,
   * צריך לבחור שחקן טוב מקבוצה שתנצח.
   */
  appliesTo: 'personal' | 'personalAndResult';
  applyToNegative: boolean;
  immunity: {
    yellowCards: boolean;
    redCards: boolean;
    minMinutes: number;
    /** תוצאות שבהן החסינות חלה */
    requireResultIn: Array<'W' | 'D' | 'L'>;
  };
  transferToViceIfMinutesBelow: number;
}

export interface ConstraintRules {
  lineupSize: number;
  maxPlayersPerTeam: number;
  benchSize: number;
  formationAllowed: string[];
  requireCaptain: boolean;
}

export interface RuleSet {
  version: number;
  leagueCode: string | null;
  personal: PersonalRules;
  resultBonus: ResultBonusRules;
  virtualGoal: VirtualGoalRules;
  captain: DubidCaptainRules;
  constraints: ConstraintRules;
}

/* -------------------------------------------------------------------
   ערכי הניקוד לפי המפרט שהוגדר למוצר:
     שער   — שוער/מגן 6 · קשר 5 · חלוץ 4
     בישול — 3
     קלין-שיט — 4 (אחיד, דורש 60 דקות)
     כרטיס — צהוב ‎-1‎ · אדום ‎-3‎
   ------------------------------------------------------------------- */
const GOAL: Record<Position, number> = { GK: 6, DEF: 6, MID: 5, FWD: 4 };
const CLEAN: Record<Position, number> = { GK: 4, DEF: 4, MID: 4, FWD: 4 };
const CONCEDED: Record<Position, number> = { GK: 0, DEF: 0, MID: 0, FWD: 0 };

/** ליגת העל הישראלית — ברירת המחדל של המוצר. */
export const IL_PREMIER: RuleSet = {
  version: 1,
  leagueCode: 'IL_PREMIER',
  personal: {
    goal: GOAL,
    assist: 3,
    cleanSheet: CLEAN,
    cleanSheetMinMinutes: 60,

    // ★ אין נקודות על עצם ההשתתפות — לפי המפרט.
    //   שימו לב לתוצאה: מגן ששיחק 90 דקות בלי קלין-שיט ובלי שער מקבל 0.
    //   אם תרצו "נקודות הופעה" (סטנדרט מקובל בפנטזי, מרכך את המזל):
    //   minutesPlayed: 1, minutes60Plus: 2  — שינוי של שורה אחת.
    minutesPlayed: 0,
    minutes60Plus: 0,

    yellowCard: -1,
    redCard: -3,

    // תוספת מעבר למפרט: שער עצמי. אפשר לאפס אם לא רוצים.
    ownGoal: -2,

    // מנוטרלים כברירת מחדל כדי להיצמד למפרט. השדות קיימים —
    // הדלקה = שינוי ערך, לא שינוי קוד.
    savesPer: 3,
    savesPoints: 0,
    penaltySaved: 0,
    penaltyMissed: 0,
    goalsConcededPer: 2,
    goalsConceded: CONCEDED,
  },
  resultBonus: { W: 4, D: 1, L: 0, requireMinutes: 0 },
  // סכימת שערי כל 11 הקבוצות לקופה אחת; כל 2 שערים = שער וירטואלי של 5 נק׳.
  virtualGoal: { goalsPerVirtual: 2, points: 5, aggregation: 'pooled' },
  captain: {
    multiplier: 3,
    appliesTo: 'personalAndResult',
    applyToNegative: false,
    immunity: {
      yellowCards: true,
      redCards: false,
      minMinutes: 60,
      requireResultIn: ['W', 'D'],
    },
    transferToViceIfMinutesBelow: 1,
  },
  constraints: {
    lineupSize: 11,
    maxPlayersPerTeam: 1,
    benchSize: 0,
    formationAllowed: ['4-3-3', '4-4-2', '3-5-2', '4-2-3-1', '5-3-2', '3-4-3'],
    requireCaptain: true,
  },
};

/**
 * דוגמה להרחבה לליגה זרה — 15 שחקנים, עד 2 מאותה קבוצה, קפטן ×2.
 * שימו לב: אין כאן שום קוד חדש. רק ערכים.
 */
export const EN_PREMIER: RuleSet = {
  ...IL_PREMIER,
  leagueCode: 'EN_PL',
  virtualGoal: { goalsPerVirtual: 3, points: 6, aggregation: 'perTeam' },
  captain: { ...IL_PREMIER.captain, multiplier: 2 },
  constraints: {
    ...IL_PREMIER.constraints,
    lineupSize: 15,
    maxPlayersPerTeam: 2,
    benchSize: 4,
  },
};

/**
 * דוביד 5 על 5 — הפורמט המהיר מהמשחק המקורי, חי עכשיו בתוך אותו מנוע.
 * 5 שחקנים (שוער + 4 שחקני שדה), תקציב 15 מיליון יורו, שחקן אחד מכל
 * קבוצה, שני מערכים בדיוק כמו במקור: 2-1-1 ו-1-2-1 (הגנה-קישור-התקפה,
 * שוער תמיד בנפרד). אותו `scoreLineup`, אותו `validateLineup` —
 * הפורמט הקטן לא צריך קוד חדש, רק ערכים אחרים.
 */
export const DUBID_5X5: RuleSet = {
  ...IL_PREMIER,
  version: 1,
  constraints: {
    ...IL_PREMIER.constraints,
    lineupSize: 5,
    maxPlayersPerTeam: 1,
    formationAllowed: ['2-1-1', '1-2-1'],
  },
};

/** תקציב 5 על 5, במיליוני יורו — אילוץ UI, לא חלק מהמנוע הגנרי. */
export const DUBID_5X5_BUDGET = 15;


export function ruleSetFromJson(
  data: Partial<RuleSet> & Record<string, unknown>,
  base: RuleSet = IL_PREMIER,
): RuleSet {
  return {
    ...base,
    ...data,
    personal: { ...base.personal, ...(data.personal ?? {}) },
    resultBonus: { ...base.resultBonus, ...(data.resultBonus ?? {}) },
    virtualGoal: { ...base.virtualGoal, ...(data.virtualGoal ?? {}) },
    captain: {
      ...base.captain,
      ...(data.captain ?? {}),
      immunity: { ...base.captain.immunity, ...(data.captain?.immunity ?? {}) },
    },
    constraints: { ...base.constraints, ...(data.constraints ?? {}) },
  } as RuleSet;
}

/* =================================================================== */
/* התאמת החוקים לכמות הקבוצות שבפועל יש בליגה                          */
/* =================================================================== */

/**
 * ★ אריתמטיקה שאי אפשר לעקוף:
 *   עם `maxPlayersPerTeam = 1`, מספר השחקנים בהרכב לא יכול לעלות על
 *   מספר הקבוצות בליגה. 11 שחקנים דורשים 11 קבוצות שונות.
 *
 * כשהדאטה חלקי (למשל 6 קבוצות מתוך 14), יש שתי אפשרויות גרועות
 * ואחת טובה:
 *   ✗ להגדיל את maxPlayersPerTeam — שובר את חוק הליבה של המשחק
 *   ✗ להשאיר 11 ולתת למשתמש להיתקע במסך חצי-ריק
 *   ✓ להריץ מצב הדגמה עם הרכב מוקטן, ולומר את זה במפורש
 *
 * הפונקציה מחזירה את החוקים המותאמים + דגל שה-UI חייב להציג.
 */
export interface ResolvedRules {
  rules: RuleSet;
  isDemo: boolean;
  teamsAvailable: number;
  teamsRequired: number;
  missingTeams: number;
}

/** מערך מתאים לגודל הרכב נתון (כולל שוער). */
export function formationForSize(size: number): string {
  const table: Record<number, string> = {
    5: '2-1-1', 6: '2-2-1', 7: '3-2-1', 8: '3-3-1',
    9: '3-3-2', 10: '4-3-2', 11: '4-3-3',
  };
  return table[size] ?? '4-3-3';
}

export function resolveRules(teamsAvailable: number, base: RuleSet = IL_PREMIER): ResolvedRules {
  const required = base.constraints.lineupSize * base.constraints.maxPlayersPerTeam;
  if (teamsAvailable >= required) {
    return { rules: base, isDemo: false, teamsAvailable, teamsRequired: required, missingTeams: 0 };
  }

  const size = Math.max(5, Math.floor(teamsAvailable / base.constraints.maxPlayersPerTeam));
  return {
    rules: {
      ...base,
      constraints: {
        ...base.constraints,
        lineupSize: size,
        formationAllowed: [formationForSize(size)],
      },
    },
    isDemo: true,
    teamsAvailable,
    teamsRequired: required,
    missingTeams: required - teamsAvailable,
  };
}
